sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs

```diff
@@ -1,151 +1,238 @@
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static Dictionary<KnownType, int> mapKnownTypesToIntegerBitSize
+            = new Dictionary<KnownType, int>
+            {
+                [KnownType.System_Int64] = 64,
+                [KnownType.System_UInt64] = 64,
+
+                [KnownType.System_Int32] = 32,
+                [KnownType.System_UInt32] = 32,
+
+                [KnownType.System_Int16] = 16,
+                [KnownType.System_UInt16] = 16,
+
+                [KnownType.System_Byte] = 8,
+                [KnownType.System_SByte] = 8
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var shiftInstances = (c.Node as MemberDeclarationSyntax)
+                        .DescendantNodes()
+                        .Select(n => FindIssues(n, c.SemanticModel))
+                        .WhereNotNull();
+
+                    var zeroShiftIssues = new List<ShiftInstance>();
+                    var linesWithShiftOperations = new HashSet<int>();
+
+                    foreach (var shift in shiftInstances)
+                    {
+                        if (shift.IsLiteralZero)
+                        {
+                            zeroShiftIssues.Add(shift);
+                            continue;
+                        }
+
+                        linesWithShiftOperations.Add(shift.Line);
+                        if (shift.Diagnostic != null)
+                        {
+                            c.ReportDiagnosticWhenActive(shift.Diagnostic);
+                        }
+                    }
+
+                    foreach (var issue in zeroShiftIssues)
+                    {
+                        if (!linesWithShiftOperations.Contains(issue.Line - 2) &&
+                            !linesWithShiftOperations.Contains(issue.Line - 1) &&
+                            !linesWithShiftOperations.Contains(issue.Line) &&
+                            !linesWithShiftOperations.Contains(issue.Line + 1) &&
+                            !linesWithShiftOperations.Contains(issue.Line + 2))
+                        {
+                            c.ReportDiagnosticWhenActive(issue.Diagnostic);
+                        }
+                    }
+                },
+                SyntaxKind.MethodDeclaration,
+                SyntaxKind.PropertyDeclaration);
+        }
+
+        private static Tuple<bool, ExpressionSyntax> GetRhsArgumentOfShiftNode(SyntaxNode node)
+        {
+            var binaryExpression = node as BinaryExpressionSyntax;
+            if (binaryExpression?.OperatorToken.IsKind(SyntaxKind.LessThanLessThanToken) ?? false)
+            {
+                return new Tuple<bool, ExpressionSyntax>(true, binaryExpression.Right);

```

Comment:
It looks like this method is returning _true_ for _less than_ and _false_ for _more than_. An enum would be easier to read.